File: mvn-advent-of-code/src/main/java/advent/twenty_sixteen/Day07InternetProtocolVersionSeven.java
package advent.twenty_sixteen;

import advent.common.DailyProblem;
import advent.common.Pair;
import advent.utilities.FileUtilities;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.function.Function;

public class Day07InternetProtocolVersionSeven implements DailyProblem<Integer, Integer> {
    private final int part1Answer;
    private final int part2Answer;

    private static boolean check(String text) {
        for (int i = 0; i != text.length() - 3; i++) {
            if (text.charAt(i) == text.charAt(i + 3) && text.charAt(i + 1) == text.charAt(i + 2) && text.charAt(i) != text.charAt(i + 1)) {
                return true;
            }
        }
        return false;
    }

    private static boolean check(List<String> text) {
        for (String t : text) {
            if (check(t)) {
                return true;
            }
        }
        return false;
    }

    private static final Function<String, Pair<List<String>, List<String>>> PARSE = new Function<String, Pair<List<String>, List<String>>>() {
        @Override
        public Pair<List<String>, List<String>> apply(String line) {
            List<String> outside = new ArrayList<>();
            List<String> inside = new ArrayList<>();
            boolean ip = true;
            int current = 0;
            while (current < line.length()) {
                if (ip) {
                    int index = line.indexOf("[", current);
                    if (index < 0) {
                        index = line.length();
                    }
                    outside.add(line.substring(current, index));
                    current = index + 1;
                } else {
                    int index = line.indexOf("]", current);
                    inside.add(line.substring(current, index));
                    current = index + 1;
                }
                ip = !ip;
            }
            return new Pair<>(outside, inside);
        }
    };

    public Day07InternetProtocolVersionSeven(String filename) {
        List<Pair<List<String>, List<String>>> addresses = FileUtilities.readLines(filename, PARSE);

        int part1Count = 0;
        int part2Count = 0;
        for (Pair<List<String>, List<String>> e : addresses) {
            if (check(e.getFirst()) && !check(e.getSecond())) {
                part1Count++;
            }

            if (check(split(e.getFirst()), split(e.getSecond()))) {
                part2Count++;
            }
        }

        this.part1Answer = part1Count;
        this.part2Answer = part2Count;
    }

    private boolean check(Set<String> firstParings, Set<String> secondParings) {
        for (String first : firstParings) {
            String search = String.format("%c%c%c", first.charAt(1), first.charAt(0), first.charAt(1));
            if (secondParings.contains(search)) {
                return true;
            }
        }
        return false;
    }

    private Set<String> split(List<String> chunks) {
        Set<String> result = new HashSet<>();
        for (String chunk : chunks) {
            for (int i = 0; i != chunk.length() - 2; i++) {
                if (chunk.charAt(i) == chunk.charAt(i + 2) && chunk.charAt(i) != chunk.charAt(i + 1)) {
                    result.add(chunk.substring(i, i + 3));
                }
            }
        }
        return result;
    }

    @Override
    public Integer getPart1Answer() {
        return part1Answer;
    }

    @Override
    public Integer getPart2Answer() {
        return part2Answer;
    }
}

File: scala-advent-of-code/src/main/scala/advent/twenty_sixteen/Day07InternetProtocolVersionSeven.scala
package advent.twenty_sixteen

import advent.common.DailyProblem
import advent.utilities.FileUtilities

import scala.annotation.tailrec

class Day07InternetProtocolVersionSeven(filename: String) extends DailyProblem[Int, Int] {

  private val input = FileUtilities.readFile(filename)

  private def part1(lines: List[String]): Int = {
    @tailrec
    def abba(str: String): Boolean = {
      if (str.size < 4) {
        false
      } else {
        (str(0) == str(3)) && (str(1) == str(2) && (str(0) != str(1))) || abba(str.tail)
      }
    }

    def check(line: String): Boolean = {
      val chunks = line.split("[\\[\\]]").foldLeft((false, false, false))((acc, chunk) => (!acc._1,
        if (!acc._1) acc._2 || abba(chunk) else acc._2,
        if (acc._1) acc._3 || abba(chunk) else acc._3))
      chunks._2 && !chunks._3
    }

    lines.foldLeft(0)((acc, line) => if (check(line)) acc + 1 else acc)
  }

  private def part2(lines: List[String]): Int = {
    @tailrec
    def aba(str: String, acc: Set[String]): Set[String] = {
      if (str.size < 3) {
        acc
      } else {
        aba(str.tail, if (str(0) == str(2) && str(0) != str(1)) acc + str.take(3) else acc)
      }
    }

    def check(line: String): Boolean = {
      val (_, outside, inside) = line.split("[\\[\\]]").foldLeft((false, Set[String](), Set[String]()))((acc, chunk) => (!acc._1,
        if (!acc._1) aba(chunk, acc._2) else acc._2,
        if (acc._1) aba(chunk, acc._3) else acc._3))

      outside.toList.foldLeft(false)((acc, aba) => if (inside.contains("" + aba(1) + aba(0) + aba(1))) true else acc)
    }

    lines.foldLeft(0)((acc, line) => if (check(line)) acc + 1 else acc)
  }

  override val part1Answer: Int = part1(input)
  override val part2Answer: Int = part2(input)
}
